src/RWS.jl

```diff
@@ -128,3 +130,115 @@ function Init(rws::RewritingSystem = RewritingSystem(),
+                  maxlenright = 0,
+                  maxoverlaplen = 0,
+                  maxwdiffs = 512,
+                  maxreducelen = 32767,
+                  rkminlen = 0,
+                  rkmineqns = 0,
+                  ordering::KBMOrderings = SHORTLEX)
+    rws = Init()
+
+    # Do not set the name
+    rws.name = ntuple(_->Cchar(0), 256)
+
+    # Set basic parameters
+    if isConfluent && !rws.resume_with_orig
+        error("System is already confluent!")
+    end
+
+    if !rws.tidyintset && tidyint > 0
+        rws.tidyint = tidyint
+        rws.tidyintset = true # This was not in the original file
+    end
+
+    if rws.inv_of != C_NULL
+        error("Input error: 'maxeqns' field must precede 'inverses' field")
+    elseif !rws.maxeqnsset && !rws.resume_with_orig && maxeqns > 16
+        rws.maxeqns = maxeqns
+        rws.maxeqnsset = true # This was not in the original file
+    end
+
+    if rws.inv_of != C_NULL
+        error("Input error: 'maxstates' field must precede 'inverses' field")
+    elseif !rws.maxstatesset && maxstates > 128
+        rws.maxstates = maxstates
+        rws.maxstatesset = true # This was not in the original file
+    end
+
+    if !rws.confnumset && confnum > 0
+        rws.confnum = confnum
+        rws.confnumset = true # This was not in the original file
+    end
+
+    rws.sorteqns = sorteqns
+
+    if rws.maxoplen == 0 && maxoplen >= 0
+        rws.sorteqns = true
+        rws.maxoplen = maxoplen
+    end
+
+    if maxlenleft > 0 && maxlenright > 0 && rws.maxlenleft == 0 && rws.maxlenright == 0
+        rws.maxlenleft = maxlenleft
+        rws.maxlenright = maxlenright
+    end
+
+    if maxoverlaplen > 0
+        rws.maxoverlaplen = maxoverlaplen
+    end
+
+    if !rws.maxwdiffset && maxwdiffs > 16
+        rws.maxwdiffs = maxwdiffs
+        rws.maxwdiffset = true # This was not in the original file
+    end
+
+    if !rws.maxreducelenset && maxreducelen > 4096
+        rws.maxreducelen = maxreducelen
+        rws.maxreducelenset = true # This was not in the original file
+    end
+
+    if rkminlen > 0 && rkmineqns > 0 && rws.rkminlen == 0 && rws.rkmineqns == 0
+        rws.rkminlen = rkminlen
+        rws.rkmineqns = rkmineqns
+        rk_on = true # This was not in the original file
+    end
+
+    if !rws.orderingset
+        rws.ordering = ordering
+        rws.orderingset = true # This was not in the original file
+    end
+
+    #=
+        mutable struct FPGroup <: AbstractFPGroup
+            gens::Vector{FPSymbol}
+            rels::Dict{FPGroupElem, FPGroupElem}
+        end
+    =#
+
+    # Generators
+    C = [""]
+    append!(C, string.(G.gens))
+    GC.@preserve C true
+    # Get the number of generators
+    rws.num_gens = size(C)[1] - 1
+    # Move the generators to the RWS. Warning. These are not processed and are assumed to be correct.
+    rws.gen_name = Base.unsafe_convert(Ptr{Ptr{Int8}}, Base.cconvert(Ptr{Ptr{Int8}}, C))
+
+    return rws
```

Comment:
my take on this thing:
```julia
C = Symbol[Symbol(""); getfield.(G.gens, :id)]
    # Get the number of generators
    rws.num_gens = length(C)
    # Move the generators to the RWS. Warning. These are not processed and are assumed to be correct.
    ref_C = Base.cconvert(Ptr{Ptr{Int8}}, C)
    rws.gen_name = Base.unsafe_convert(Ptr{Ptr{Int8}}, ref_C)

    return rws, ref_C
```
then we can do:
```julia
G = let F = FreeGroup(["a", "b", "c"])
    a,b,c = Groups.gens(F)
    rels = [a^7, b^3, c*a*b]
    F/rels
end

let (rws, stuff_to_keep_alive) = KBmag.BuildRWS(G)

    first_gen = unsafe_load(rws.gen_name, 2)
    @show unsafe_string(first_gen)

    second_gen = unsafe_load(rws.gen_name, 3)
    @show unsafe_string(second_gen)

    third_gen = unsafe_load(rws.gen_name, 4)
    @show unsafe_string(third_gen)
end
```
```
unsafe_string(first_gen) = "a"
unsafe_string(second_gen) = "b"
unsafe_string(third_gen) = "c"
```

`GC.@preserve` is used as `GC@preserve a b c function_call_involving(pointer(a), pointer(b), pointer(c),...)`, or
```julia
res = GC.@preserve a b c begin
    ptr_a = pointer(a)
    ptr_b = ...
    ...
    function_call_involving(ptr_a, ptr_b, ...)
end
```

e.g. https://github.com/JuliaOpt/SCS.jl/blob/bfc756ea4498c89c263f14cdb02dcadd063d17fa/src/c_wrapper.jl#L69

We need to store a `Ptr{Ptr{Gen}}` (a pointer to shifted vector gens) in a `rws`, so one way of keeping it from GC is to keep it alive by passing it along. The other option is to embed it into the struct itself (see https://github.com/JuliaOpt/SCS.jl/blob/bfc756ea4498c89c263f14cdb02dcadd063d17fa/src/types.jl#L20), we will explore both options.